File: JDK11/opentraceability/src/main/java/opentraceability/models/events/transactionevent.java
package opentraceability.models.events;

import opentraceability.interfaces.IEvent;
import opentraceability.models.attributes.*;
import opentraceability.models.identifiers.EPC;
import opentraceability.models.product.EventProduct;
import opentraceability.models.product.EventProductType;
import opentraceability.utility.EPCISVersion;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;

public abstract class TransactionEvent extends EventBase implements IEvent {
    @OpenTraceabilityAttribute("", "parentID", 8)
    public EPC parentID;

    @OpenTraceabilityProductsAttribute(value = {"extension/quantityList"}, versionV1 = EPCISVersion.V1, productTypeV1 = EventProductType.Reference, priorityV1 = 20, listTypeV1 = OpenTraceabilityProductsListType.QuantityList, valueV2 = "quantityList", productTypeV2 = EventProductType.Reference, priorityV2 = 15, listTypeV2 = OpenTraceabilityProductsListType.QuantityList, valueV3 = "epcList", productTypeV3 = EventProductType.Reference, priorityV3 = 9, listTypeV3 = OpenTraceabilityProductsListType.EPCList)
    public List<EventProduct> referenceProducts = new ArrayList<>();

    @OpenTraceabilityAttribute("", "action", 10)
    public EventAction action = null;

    @OpenTraceabilityAttribute("", "bizStep", 11)
    public URI businessStep = null;

    @OpenTraceabilityAttribute("", "disposition", 12)
    public URI disposition = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "readPoint", 13)
    public EventReadPoint readPoint = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "bizLocation", 14)
    public EventLocation location = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("bizTransaction")
    @OpenTraceabilityAttribute("", "bizTransactionList", 7)
    public List<EventBusinessTransaction> bizTransactionList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("source")
    @OpenTraceabilityAttribute(value = {"sourceList", "extension/sourceList"}, versionV2 = EPCISVersion.V2, priorityV2 = 16, productTypeV2 = EventProductType.Reference, listTypeV2 = OpenTraceabilityProductsListType.SensorElementList, valueV1 = "bizTransactionList", productTypeV1 = EventProductType.Reference, priorityV1 = 21, listTypeV1 = OpenTraceabilityProductsListType.SensorElementList)
    public List<EventSource> sourceList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("destination")
    @OpenTraceabilityAttribute("","destinationList", 17, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","extension/destinationList", 22, EPCISVersion.V1)
    public List<EventDestination> destinationList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("sensorElement")
    @OpenTraceabilityAttribute(value = {"sensorElementList", "extension/sensorElementList"}, versionV2 = EPCISVersion.V2, priorityV2 = 18, productTypeV2 = EventProductType.Reference, listTypeV2 = OpenTraceabilityProductsListType.SensorElementList, valueV1 = "bizTransactionList", productTypeV1 = EventProductType.Reference, priorityV1 = 18, listTypeV1 = OpenTraceabilityProductsListType.SensorElementList)
    public List<SensorElement> sensorElementList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(value = {"persistentDisposition", "extension/persistentDisposition"}, versionV2 = EPCISVersion.V2, priorityV2 = 19, productTypeV2 = EventProductType.Extension, listTypeV2 = OpenTraceabilityProductsListType.QuantityList, valueV1 = "bizTransactionList", productTypeV1 = EventProductType.Extension, priorityV1 = 19, listTypeV1 = OpenTraceabilityProductsListType.EPCList)
    public PersistentDisposition persistentDisposition = null;

    public EventILMD ILMD = null;

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("", "type", 0)
    public EventType eventType;

    public List<EventProduct> products = new ArrayList<>();

    @Override
    public void addProduct(EventProduct product) throws Exception {
        if (product.Type == EventProductType.Parent) {
            if (product.Quantity != null) {
                throw new Exception("Parents do not support quantity.");
            }
            this.parentID = product.EPC;
        } else if (product.Type == EventProductType.Reference) {
            this.referenceProducts.add(product);
        } else {
            throw new Exception("Transaction event only supports references and parents.");
        }
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/models/events/transactionevent.kt
package models.events

import interfaces.IEvent
import java.util.*
import models.identifiers.EPC
import utility.attributes.*
import java.net.URI

abstract class TransactionEvent : EventBase(), IEvent {

    @OpenTraceabilityAttribute("","parentID", 8)
    var ParentID: EPC? = null

    @OpenTraceabilityProductsAttribute("extension/quantityList", EPCISVersion.V1, EventProductType.Reference, 20, OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute("quantityList", EPCISVersion.V2, EventProductType.Reference, 15, OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute("epcList",EPCISVersion.V2, EventProductType.Reference, 9, OpenTraceabilityProductsListType.EPCList)
    var ReferenceProducts: ArrayList<EventProduct> = ArrayList<EventProduct>()

    @OpenTraceabilityAttribute("","action", 10)
    override var Action: EventAction? = null

    @OpenTraceabilityAttribute("","bizStep", 11)
    override var BusinessStep: URI? = null

    @OpenTraceabilityAttribute("","disposition", 12)
    override var Disposition: URI? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","readPoint", 13)
    override var ReadPoint: EventReadPoint? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","bizLocation", 14)
    override var Location: EventLocation? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("bizTransaction")
    @OpenTraceabilityAttribute("","bizTransactionList", 7)
    override var BizTransactionList: MutableList<EventBusinessTransaction> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("source")
    @OpenTraceabilityAttribute("","sourceList", 16, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","extension/sourceList", 21, EPCISVersion.V1)
    override var SourceList: MutableList<EventSource> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("destination")
    @OpenTraceabilityAttribute("","destinationList", 17, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","extension/destinationList", 22, EPCISVersion.V1)
    override var DestinationList: MutableList<EventDestination> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute("sensorElement")
    @OpenTraceabilityAttribute("","sensorElementList", 18, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","extension/sensorElementList",18, EPCISVersion.V1)
    override var SensorElementList: MutableList<SensorElement> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","persistentDisposition", 19, EPCISVersion.V2)
    @OpenTraceabilityAttribute("","extension/persistentDisposition", 19, EPCISVersion.V1)
    override var PersistentDisposition: PersistentDisposition? = null

    var ILMD: EventILMD? = null


    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("","type", 0)
    override lateinit var EventType: EventType

    override var Products: MutableList<EventProduct> = mutableListOf()

    override fun AddProduct(product: EventProduct){
        if (product.Type == EventProductType.Parent)
        {
            if (product.Quantity != null)
            {
                throw  Exception("Parents do not support quantity.");
            }
            this.ParentID = product.EPC;
        }
        else if (product.Type == EventProductType.Reference)
        {
            this.ReferenceProducts.add(product);
        }
        else
        {
            throw Exception("Transaction event only supports references and parents.");
        }
    }
}
